Common/AR/source/Android/ArCoreWrapper.cpp

```diff
@@ -0,0 +1,933 @@
+  return QColor();
+}
+
+/*!
+  \internal
+ */
+void ArCoreWrapper::setPointCloudColor(const QColor& pointCloudColor)
+{
+  if (pointCloudColor.isValid())
+  {
+    if (!m_arCorePointCloudRenderer)
+    {
+      m_arCorePointCloudRenderer.reset(new ArCorePointCloudRenderer(this));
+      m_arCorePointCloudRenderer->initGL();
+    }
+
+    m_arCorePointCloudRenderer->setPointCloudColor(pointCloudColor);
+  }
+  else
+  {
+    m_arCorePointCloudRenderer.reset();
+  }
+}
+
+/*!
+  \internal
+ */
+int ArCoreWrapper::pointCloudSize() const
+{
+  if (m_arCorePointCloudRenderer)
+    return m_arCorePointCloudRenderer->pointCloudSize();
+  return -1;
+}
+
+/*!
+  \internal
+ */
+void ArCoreWrapper::setPointCloudSize(int pointCloudSize)
+{
+  if (pointCloudSize > 0)
+  {
+    if (!m_arCorePointCloudRenderer)
+    {
+      m_arCorePointCloudRenderer.reset(new ArCorePointCloudRenderer(this));
+      m_arCorePointCloudRenderer->initGL();
+    }
+
+    m_arCorePointCloudRenderer->setPointCloudSize(pointCloudSize);
+  }
+  else
+  {
+    m_arCorePointCloudRenderer.reset();
+  }
+}
+
+/*!
+  \internal
+ */
+QColor ArCoreWrapper::planeColor() const
+{
+  if (m_arCorePlaneRenderer)
+    return m_arCorePlaneRenderer->planeColor();
+  return QColor();
+}
+
+/*!
+  \internal
+ */
+void ArCoreWrapper::setPlaneColor(const QColor& planeColor)
+{
+  if (planeColor.isValid())
+  {
+    if (!m_arCorePlaneRenderer)
+    {
+      m_arCorePlaneRenderer.reset(new ArCorePlaneRenderer(this));
+      m_arCorePlaneRenderer->initGL();
+    }
+
+    m_arCorePlaneRenderer->setPlaneColor(planeColor);
+  }
+  else
+  {
+    m_arCorePlaneRenderer.reset();
+  }
+}
+
+/*!
+  \internal
+  Methods for point cloud data.
+  Be careful with the concurrent access and lifetime: the data are initialized and
+  released in the main thread and used in the GL thread.
+ */
+void ArCoreWrapper::planeListData(int32_t& size)
+{
+  if (!m_arSession || m_arPlaneList)
+    return;
+
+  ArTrackableList_create(m_arSession, &m_arPlaneList);
+  if (m_arPlaneList == nullptr)
+    return;
+
+  ArTrackableType planeTrackedType = AR_TRACKABLE_PLANE;
+  ArSession_getAllTrackables(m_arSession, planeTrackedType, m_arPlaneList);
+
+  ArTrackableList_getSize(m_arSession, m_arPlaneList, &size);
+}
+
+/*!
+  \internal
+ */
+void ArCoreWrapper::releasePlaneListData()
+{
+  if (!m_arPlaneList)
+    return;
+
+  ArTrackableList_destroy(m_arPlaneList);
+  m_arPlaneList = nullptr;
+}
+
+/*!
+  \internal
+ */
+bool ArCoreWrapper::planeData(int32_t index, QMatrix4x4& mvp, std::vector<float>& vertices)
+{
+  ArTrackableList_acquireItem(m_arSession, m_arPlaneList, index, &m_arTrackable);
+  ArPlane* arPlane = ArAsPlane(m_arTrackable);
+
+  ArTrackingState outTrackingState = AR_TRACKING_STATE_STOPPED;
+  ArTrackable_getTrackingState(m_arSession, m_arTrackable, &outTrackingState);
+
+  ArPlane* subsumePlane = nullptr;
+  ArPlane_acquireSubsumedBy(m_arSession, arPlane, &subsumePlane);
+  if (subsumePlane != nullptr)
```

Comment:
I think in some cases it may not be obvious you're dealing with a pointer. In this case it's obvious.

```suggestion
  if (subsumePlane)
```